kafka-daq-report/src/main.rs

```diff
@@ -228,36 +242,55 @@ async fn poll_kafka_msg(consumer: StreamConsumer, shared_data: SharedData) {
-                                        num_samples_in_first_channel,
-                                        is_num_samples_identical,
-                                    ));
+                                let id = data.digitizer_id();
+                                {
+                                    let mut logged_data = shared_data.lock().unwrap();
+                                    logged_data
+                                        .entry(id)
+                                        .and_modify(|d| {
+                                            d.num_msg_received += 1;
+                                            d.recent_msg_count += 1;
+
+                                            d.last_msg_timestamp = timestamp;
+
+                                            d.message_rate = d.recent_msg_count as f64
+                                                / recent_message_lifetime as f64;
+
+                                            let num_channels = match data.channels() {
+                                                Some(c) => c.len(),
+                                                None => 0,
+                                            };
+                                            if !d.has_num_channels_changed {
+                                                d.has_num_channels_changed =
+                                                    num_channels != d.num_channels_present;
+                                            }
+                                            d.num_channels_present = num_channels;
+                                            if !d.has_num_channels_changed {
+                                                d.has_num_samples_changed =
+                                                    num_samples_in_first_channel
+                                                        != d.num_samples_in_first_channel;
+                                            }
+                                            d.num_samples_in_first_channel =
+                                                num_samples_in_first_channel;
+                                            d.is_num_samples_identical = is_num_samples_identical;
+                                        })
+                                        .or_insert(DigitiserData::new(
+                                            timestamp,
+                                            frame_number,
+                                            num_channels_present,
+                                            num_samples_in_first_channel,
+                                            is_num_samples_identical,
+                                        ));
+                                };
+
+                                let shared_data = Arc::clone(&shared_data);
+                                thread::spawn(move || {
+                                    thread::sleep(Duration::from_secs(recent_message_lifetime));
```

Comment:
This appears to only compute the message rate after a given delay of a message being received.
This is not the intended behaviour, the message rate should be calculated at a constant rate (defined by `recent_message_lifetime` here) regardless of what messages are being received.
This is why you are seeing issues when no more messages are received.

Reply:
I set the code up to track the number of messages detected in the last `recent_message_lifetime` seconds, increasing the count whenever a message is detected and decreasing the count after the set period of time. This way the `message_rate` can be calculated dynamically every frame. I'm not sure why this approach doesn't work. Surely the message rate must depend on the messages being received?

Reply:
The two issues with this are:
- You do not have a constant sampling window, so the value displayed may fluctuate needlessly.
- It does not work if the message rate is below the sampling rate (which should not typically happen, but this is a diagnostics tool, so should be able to cope with things not working as intended).
- (a more technical point is that spawning a task or (especially) thread for each received message is needlessly expensive)

As number of messages received is already being recorded, all that you need to do is evey n seconds calculate the difference in this number since the last iteration, divide it by n and you have the message rate in Hz.

Reply:
Ok, I will use that approach instead. That should mean that the `message_rate` is only updated every `recent_message_lifetime` seconds but I assume that is not an issue?

Reply:
> Ok, I will use that approach instead. That should mean that the `message_rate` is only updated every `recent_message_lifetime` seconds but I assume that is not an issue?

That is the intention, although the parameter should be renamed to better reflect this.